server/src/main/java/org/elasticsearch/cluster/coordination/ElectionScheduler.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Setting.Property;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Random;
+import java.util.function.Function;
+
+public abstract class ElectionScheduler extends AbstractComponent {
+
+    /*
+     * It's provably impossible to guarantee that any leader election algorithm ever elects a leader, but they generally work (with
+     * probability that approaches 1 over time) as long as elections occur sufficiently infrequently, compared to the time it takes to send
+     * a message to another node and receive a response back. We do not know the round-trip latency here, but we can approximate it by
+     * attempting elections randomly at reasonably high frequency and backing off (linearly) until one of them succeeds. We also place an
+     * upper bound on the backoff so that if elections are failing due to a network partition that lasts for a long time then when the
+     * partition heals there is an election attempt reasonably quickly.
+     */
+
+    // bounds on the time between election attempts
+    private static final String ELECTION_MIN_RETRY_INTERVAL_SETTING_KEY = "discovery.election.min_retry_interval";
+    private static final String ELECTION_MAX_RETRY_INTERVAL_SETTING_KEY = "discovery.election.max_retry_interval";
```

Comment:
I think we should go with naming that's more established in the literature, and use something like `min_election_timeout` and `max_election_timeout`

Reply:
Ok. I don't particularly like the "timeout" nomenclature - "timeout" suggests how long some process has before it's considered to have failed, but here we're counting down until something starts. But I can live with it.